src/components/lib/crt/crt.c

```diff
@@ -167,6 +217,74 @@ crt_comp_create_with(struct crt_comp *c, char *name, compid_t id, struct crt_com
 	cos_compinfo_init(cos_compinfo_get(c->comp_res),
 			  r->ptc, r->ctc, r->compc, r->heap_ptr, r->captbl_frontier,
 			  cos_compinfo_get(cos_defcompinfo_curr_get()));
+	return 0;
+}
+
+/** 
+ * Create the component from the checkpoint
+ * c is the new component
+ * chkpt is the checkpoint used to create c
+ * name is c's name
+ * id is c's component id
+ */
+int 
+crt_comp_create_from(struct crt_comp *c, char *name, compid_t id, struct crt_chkpt *chkpt)
+{
+	struct cos_compinfo *ci, *root_ci;
+	struct cos_component_information *comp_info;
+	unsigned long info_offset;
+	size_t  ro_sz,   rw_sz, data_sz, bss_sz;
+	char   *ro_src, *data_src, *mem;
+	int     ret;
+	vaddr_t	info = chkpt->c->info;
+
+	assert(c && name);
+
+	if (crt_comp_init(c, name, id, NULL, chkpt->c->info)) BUG();
+	ci      = cos_compinfo_get(c->comp_res);
+	root_ci = cos_compinfo_get(cos_defcompinfo_curr_get());
+
+	// overwrite
+	c->entry_addr = chkpt->c->entry_addr;
+	c->ro_addr = chkpt->c->ro_addr;
+	c->rw_addr = chkpt->c->rw_addr;
+
+	//re-work the sinvs with the right IDs
+	memcpy(c->sinvs, chkpt->c->sinvs, sizeof(c->sinvs));
+	c->n_sinvs = chkpt->c->n_sinvs;
+	for(u32_t i = 0; i < c->n_sinvs; i++) {
+		//if the client was the old component, update id in new
+		if(chkpt->c->sinvs[i]->client->id == chkpt->c->id) {
+			c->sinvs[i]->client = c;
```

Comment:
sinvs from a component should *only have* `->client->id == chkpt->c->id`. And none of the servers should have its id. If all of this cannot be replaced with the following, there might be something very strange going on:

```c
struct crt_sinv inv = &chkpt->c->sinvs[i];

assert(inv->client->id == chkpt->c->id);
c->sinvs[i]->client = c;
assert(inv->server->id != chkpt->c->id);
```
If I'm wrong, then either 1. the sinv saving code is doing something strange, or 2. I'm missing something here.

Reply:
ok, the server part is never getting called in the test I have, so I will replace with the above